network/src/peers/peer/cipher.rs

```diff
@@ -0,0 +1,105 @@
+// Copyright (C) 2019-2021 Aleo Systems Inc.
+// This file is part of the snarkOS library.
+
+// The snarkOS library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The snarkOS library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the snarkOS library. If not, see <https://www.gnu.org/licenses/>.
+
+use std::convert::TryInto;
+
+use snow::TransportState;
+#[cfg(test)]
+use tokio::io::{AsyncRead, AsyncReadExt};
+use tokio::io::{AsyncWrite, AsyncWriteExt};
+
+use crate::NetworkError;
+
+pub struct Cipher {
+    state: TransportState,
+    buffer: Box<[u8]>,
+    noise_buffer: Box<[u8]>,
+}
+
+impl Cipher {
+    pub fn new(state: TransportState, buffer: Box<[u8]>, noise_buffer: Box<[u8]>) -> Self {
+        assert_eq!(buffer.len(), crate::MAX_MESSAGE_SIZE + 4096);
```

Comment:
what's the surplus size for?

Reply:
so we can fit the 4-byte header (we seemed to limit length of packet not including the header). did some extra because someone asked for it IIRC, or just a clean page size. don't recall. if you care, we can probably tune it. might have been related to noise padding bug or something like that.

Reply:
If we intend to keep this padding, `MAX_MESSAGE_SIZE` should be defined as an equation, e.g.:
```rust
/// Message size composed of size header and message.
const MAX_MESSAGE_SIZE: usize = 12345 + 4096;
```